constraint-solver/src/solver/backtracking.rs

```diff
@@ -0,0 +1,159 @@
+            "Found {} sets of variables with few possible assignments. Checking each set...",
+            variable_sets.len()
+        );
+
+        let unique_assignments = variable_sets
+            .iter()
+            .map(|assignment_candidates| self.find_unique_assignment(assignment_candidates))
+            // Might error out of a contradiction was found.
+            .collect::<Result<Vec<_>, _>>()?
+            .into_iter()
+            // Might return None if the assignment is not unique.
+            .flatten()
+            .collect::<Vec<_>>();
+
+        log::debug!(
+            "{} variable sets with unique assignments found",
+            unique_assignments.len()
+        );
+
+        let mut result = BTreeMap::new();
+        for (variable, value) in unique_assignments.iter().flatten() {
+            if let Some(old_value) = result.insert(variable.clone(), *value) {
+                if old_value != *value {
+                    // Two assignments contradict each other.
+                    return Err(Error::BacktrackingError);
+                }
+            }
+        }
+
+        log::debug!("Total assignments: {}", result.len());
+        for (variable, value) in &result {
+            log::trace!("  {variable} = {value}");
+        }
+
+        Ok(result)
+    }
+
+    /// Returns all unique sets of variables that appear together in an identity
+    /// (either in an algebraic constraint or in the same field of a bus interaction),
+    /// IF the number of possible assignments is less than `MAX_BACKTRACKING_WIDTH`.
+    fn get_brute_force_candidates(&self) -> Vec<BTreeSet<V>> {
+        self.solver
+            .constraint_system
+            .iter()
+            .map(|expression| {
+                expression
+                    .referenced_variables()
+                    .cloned()
+                    .collect::<BTreeSet<_>>()
+            })
+            .unique()
+            .filter(|variables| !variables.is_empty())
+            .filter(|variables| self.has_few_possible_assignments(variables.iter().cloned()))
+            .collect()
+    }
+
+    /// Returns true if the given range constraints allow for at most `MAX_BACKTRACKING_WIDTH``
+    /// possible assignments for the given variables.
+    fn has_few_possible_assignments(&self, variables: impl Iterator<Item = V>) -> bool {
+        self.range_constraints(variables)
+            .iter()
+            .map(|rc| rc.range_width().try_into_u64())
+            .collect::<Option<Vec<_>>>()
+            .and_then(|widths| widths.iter().try_fold(1u64, |acc, &x| acc.checked_mul(x)))
+            .map(|total_width| total_width < MAX_BACKTRACKING_WIDTH)
+            .unwrap_or(false)
+    }
+
+    /// Returns the range constraints for the given variables.
+    fn range_constraints(&self, variables: impl Iterator<Item = V>) -> Vec<RangeConstraint<T>> {
+        variables
+            .map(|v| self.solver.range_constraints.get(&v))
+            .collect()
+    }
```

Comment:
```suggestion
    fn range_constraints(&self, variables: impl Iterator<Item = &V>) -> Vec<RangeConstraint<T>> {
        variables
            .map(|v| self.solver.range_constraints.get(v))
            .collect()
    }
```

Reply:
Also every use of this function immediately calls `iter()` on the return value. But if you remove the `collect()` then maybe we don't even need this as its own function?

Reply:
True! Removed the function.